File: packages/ui-swiftui/Sources/KinetixUI/DataTable.swift
//
// DataTable.swift — KinetixDataTable.
//
// Mirrors packages/ui/src/components/data-table.tsx (a @tanstack/
// react-table wrapper over the KinetixUI Table). No @tanstack equivalent,
// so sort + pagination state is hand-held here, but the rendering reuses
// KinetixTable / KinetixTableRow / KinetixTableCell and KinetixButton
// outright. A `KinetixDataColumn` gives a header, a `String` cell mapper,
// an optional `width`, and an optional `sortKey` (nil ⇒ not sortable;
// sorts lexicographically on the returned string — zero-pad numbers).
// Tapping a sortable header cycles asc → desc.
//

import SwiftUI

public struct KinetixDataColumn<Row>: Identifiable {
    public let id = UUID()
    public let header: String
    public let width: CGFloat?
    public let sortKey: ((Row) -> String)?
    public let cell: (Row) -> String

    public init(
        header: String,
        width: CGFloat? = nil,
        sortKey: ((Row) -> String)? = nil,
        cell: @escaping (Row) -> String
    ) {
        self.header = header
        self.width = width
        self.sortKey = sortKey
        self.cell = cell
    }
}

public struct KinetixDataTable<Row: Identifiable>: View {
    @Environment(\.kinetixColors) private var colors

    private let columns: [KinetixDataColumn<Row>]
    private let rows: [Row]
    private let pageSize: Int

    @State private var sortColumn: UUID?
    @State private var sortAscending = true
    @State private var page = 0

    public init(columns: [KinetixDataColumn<Row>], rows: [Row], pageSize: Int = 10) {
        self.columns = columns
        self.rows = rows
        self.pageSize = pageSize
    }

    private var sortedRows: [Row] {
        guard
            let sortColumn,
            let col = columns.first(where: { $0.id == sortColumn }),
            let key = col.sortKey
        else { return rows }
        let ascending = rows.sorted { key($0) < key($1) }
        return sortAscending ? ascending : ascending.reversed()
    }

    private var pageCount: Int {
        max(1, Int(ceil(Double(rows.count) / Double(pageSize))))
    }

    private var pageRows: [Row] {
        let start = page * pageSize
        let all = sortedRows
        guard start < all.count else { return [] }
        return Array(all[start ..< min(start + pageSize, all.count)])
    }

    private func toggleSort(_ col: KinetixDataColumn<Row>) {
        guard col.sortKey != nil else { return }
        if sortColumn == col.id {
            sortAscending.toggle()
        } else {
            sortColumn = col.id
            sortAscending = true
        }
        page = 0
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            KinetixTable {
                KinetixTableRow(isHeader: true) {
                    ForEach(columns) { col in
                        Button {
                            toggleSort(col)
                        } label: {
                            HStack(spacing: 4) {
                                Text(col.header)
                                    .font(.system(size: 14, weight: .medium))
                                if sortColumn == col.id {
                                    Image(systemName: sortAscending ? "chevron.up" : "chevron.down")
                                        .font(.system(size: 10))
                                }
                            }
                            .foregroundStyle(colors.mutedForeground)
                            .frame(maxWidth: col.width == nil ? .infinity : nil, alignment: .leading)
                            .frame(width: col.width, alignment: .leading)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 8)
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                        .disabled(col.sortKey == nil)
                    }
                }
                ForEach(pageRows) { row in
                    KinetixTableRow {
                        ForEach(columns) { col in
                            KinetixTableCell(width: col.width) {
                                Text(col.cell(row))
                                    .font(.system(size: 14))
                                    .foregroundStyle(colors.foreground)
                            }
                        }
                    }
                }
            }

            HStack {
                Text("Page \(page + 1) of \(pageCount)")
                    .font(.system(size: 13))
                    .foregroundStyle(colors.mutedForeground)
                Spacer()
                KinetixButton(variant: .outline, size: .sm, action: { if page > 0 { page -= 1 } }) {
                    Text("Previous")
                }
                KinetixButton(variant: .outline, size: .sm, action: { if page < pageCount - 1 { page += 1 } }) {
                    Text("Next")
                }
            }
        }
    }
}

File: packages/ui-flutter/lib/src/data_table.dart
import 'package:flutter/material.dart';

import 'button.dart';
import 'table.dart';
import 'app_text.dart';
import 'theme.dart';

class KinetixDataColumn<T> {
  const KinetixDataColumn({
    required this.header,
    required this.cell,
    this.width,
    this.sortKey,
  });

  final String header;
  final String Function(T row) cell;
  final double? width;

  /// Returns a string sort key (compared lexicographically — zero-pad
  /// numbers). `null` ⇒ the column isn't sortable.
  final String Function(T row)? sortKey;
}

/// Mirrors `packages/ui/src/components/data-table.tsx`. Hand-held sort +
/// pagination, reusing [KinetixTable] / [KinetixTableRow] / [KinetixButton]
/// outright. Tapping a sortable header cycles asc → desc.
class KinetixDataTable<T> extends StatefulWidget {
  const KinetixDataTable({
    super.key,
    required this.columns,
    required this.rows,
    this.pageSize = 10,
  });

  final List<KinetixDataColumn<T>> columns;
  final List<T> rows;
  final int pageSize;

  @override
  State<KinetixDataTable<T>> createState() => _KinetixDataTableState<T>();
}

class _KinetixDataTableState<T> extends State<KinetixDataTable<T>> {
  int? _sortColumn;
  bool _ascending = true;
  int _page = 0;

  List<T> get _sorted {
    final col = _sortColumn == null ? null : widget.columns[_sortColumn!];
    final key = col?.sortKey;
    if (key == null) return widget.rows;
    final sorted = [...widget.rows]..sort((a, b) => key(a).compareTo(key(b)));
    return _ascending ? sorted : sorted.reversed.toList();
  }

  int get _pageCount {
    final n = (widget.rows.length / widget.pageSize).ceil();
    return n < 1 ? 1 : n;
  }

  List<T> get _pageRows {
    final all = _sorted;
    final start = _page * widget.pageSize;
    if (start >= all.length) return const [];
    return all.sublist(start, (start + widget.pageSize).clamp(0, all.length));
  }

  void _toggleSort(int i) {
    if (widget.columns[i].sortKey == null) return;
    setState(() {
      if (_sortColumn == i) {
        _ascending = !_ascending;
      } else {
        _sortColumn = i;
        _ascending = true;
      }
      _page = 0;
    });
  }

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        KinetixTable(
          children: [
            KinetixTableRow(
              isHeader: true,
              cells: [
                for (var i = 0; i < widget.columns.length; i++)
                  _headerCell(context, i),
              ],
            ),
            for (final row in _pageRows)
              KinetixTableRow(
                cells: [
                  for (final col in widget.columns)
                    KinetixTableCell(
                      width: col.width,
                      child: Text(col.cell(row), style: AppText.bodyMd.copyWith(color: c.foreground)),
                    ),
                ],
              ),
          ],
        ),
        const SizedBox(height: 12),
        Row(
          children: [
            Text(
              'Page ${_page + 1} of $_pageCount',
              style: TextStyle(fontSize: 13, color: c.mutedForeground),
            ),
            const Spacer(),
            KinetixButton(
              variant: KinetixButtonVariant.outline,
              size: KinetixButtonSize.sm,
              onPressed: _page > 0 ? () => setState(() => _page -= 1) : null,
              child: const Text('Previous'),
            ),
            const SizedBox(width: 8),
            KinetixButton(
              variant: KinetixButtonVariant.outline,
              size: KinetixButtonSize.sm,
              onPressed: _page < _pageCount - 1 ? () => setState(() => _page += 1) : null,
              child: const Text('Next'),
            ),
          ],
        ),
      ],
    );
  }

  Widget _headerCell(BuildContext context, int i) {
    final c = KinetixTheme.of(context);
    final col = widget.columns[i];
    final Widget label = GestureDetector(
      behavior: HitTestBehavior.opaque,
      onTap: () => _toggleSort(i),
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 8),
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text(
              col.header,
              style: AppText.labelLg.copyWith(color: c.mutedForeground),
            ),
            if (_sortColumn == i) ...[
              const SizedBox(width: 4),
              Icon(_ascending ? Icons.arrow_upward : Icons.arrow_downward, size: 12, color: c.mutedForeground),
            ],
          ],
        ),
      ),
    );
    return col.width == null ? Expanded(child: label) : SizedBox(width: col.width, child: label);
  }
}
